applications/MeshingApplication/custom_utilities/tet10_refinement_utility.h

```diff
@@ -0,0 +1,499 @@
+      {
+	  rElements.push_back(*(it_new.base()));
+      }
+
+      /* All of the elements to be erased are at the end */
+      rElements.Sort();
+
+      /* Now remove all of the "old" elements */
+      rElements.erase(this_model_part.Elements().end() - to_be_deleted, this_model_part.Elements().end());
+
+      std::cout << "NEW NUMBER ELEMENTS: " << rElements.size() << std::endl;
+
+
+      // Now update the elements in SubModelParts
+      if (NewElements.size() > 0)
+      {
+          for (ModelPart::SubModelPartIterator iSubModelPart = this_model_part.SubModelPartsBegin();
+                  iSubModelPart != this_model_part.SubModelPartsEnd(); iSubModelPart++)
+          {
+              to_be_deleted = 0;
+              NewElements.clear();
+
+              // Create list of new elements in SubModelPart
+              // Count how many elements will be removed
+              for (ModelPart::ElementIterator iElem = iSubModelPart->ElementsBegin();
+                      iElem != iSubModelPart->ElementsEnd(); iElem++)
+              {
+                  if( iElem->GetValue(SPLIT_ELEMENT) )
+                  {
+                      to_be_deleted++;
+                      GlobalPointersVector< Element >& rChildElements = iElem->GetValue(NEIGHBOUR_ELEMENTS);
+
+                      for ( auto iChild = rChildElements.ptr_begin();
+                              iChild != rChildElements.ptr_end(); iChild++ )
+                      {
+                          NewElements.push_back((*iChild)->shared_from_this());
+                      }
+                  }
+              }
+
+              // Add new elements to SubModelPart
+              iSubModelPart->Elements().reserve( iSubModelPart->Elements().size() + NewElements.size() );
+              for (PointerVector< Element >::iterator it_new = NewElements.begin();
+                      it_new != NewElements.end(); it_new++)
+              {
+                  iSubModelPart->Elements().push_back(*(it_new.base()));
+              }
+
+              // Delete old elements
+              iSubModelPart->Elements().Sort();
+              iSubModelPart->Elements().erase(iSubModelPart->Elements().end() - to_be_deleted, iSubModelPart->Elements().end());
+
+              KRATOS_WATCH(iSubModelPart->Info());
+              KRATOS_WATCH(to_be_deleted);
+              KRATOS_WATCH(iSubModelPart->Elements().size());
+              KRATOS_WATCH(this_model_part.Elements().size());
+          }
+      }
+    }
+
+    /**
+    * Remove the old conditions and creates new ones
+    * @param Coord: The coordinates of the nodes of the geometry
+    * @return this_model_part: The model part of the model (it is the input too)
+    */
+
+    void EraseOldConditionsAndCreateNew(
+            ModelPart& this_model_part,
+            const compressed_matrix<int>& Coord
+            ) override
+    {
+        KRATOS_TRY;
```

Comment:
If we template this methods (that are called from the base class for any geometry), the implementation in the base class of the mesh refinement (and all its derived classes) should also be changed according to the new implentation. Furthermore, since we only treat Elements and Conditions, in my opinion  templating this two classes wouldn't make the difference worth the re-implementation.